src/Zigzag_persistence/include/gudhi/filtered_zigzag_persistence.h

```diff
@@ -0,0 +1,470 @@
+    if (itBirth == filtrationValues_.end() || itBirth->first > birthKey) {
+      --itBirth;
+    }
+    // it points to the rightmost z such that z <= x
+
+    auto itDeath =  //
+        std::lower_bound(
+            filtrationValues_.begin(), filtrationValues_.end(),
+            std::pair<internal_key, filtration_value>(deathKey, std::numeric_limits<filtration_value>::infinity()),
+            [](std::pair<internal_key, filtration_value> p1, std::pair<internal_key, filtration_value> p2) {
+              return p1.first < p2.first;
+            });
+    if (itDeath == filtrationValues_.end() || itDeath->first > deathKey) {
+      --itDeath;
+    }
+
+    return std::make_pair(itBirth->second, itDeath->second);
+  }
+
+  /**
+   * @brief Returns the current persistence diagram.
+   *
+   * @param shortestInterval Threshold. Every bar shorter than the given value will be ignored. Default value: 0.
+   * @param includeInfiniteBars If set to true, infinite bars are included in the diagram. Default value: false.
+   * @return A vector of pairs of filtration values representing the persistence diagram.
+   */
+  std::vector<Filtration_value_interval> get_persistence_diagram(filtration_value shortestInterval = 0.,
+                                                                 bool includeInfiniteBars = false) {
+    std::vector<Filtration_value_interval> diag = _get_persistence_diagram(shortestInterval);
+
+    if (includeInfiniteBars) {
+      _retrieve_infinite_bars(diag);
+    }
+
+    return diag;
+  }
+
+ private:
+  std::unordered_map<face_key, internal_key> handleToKey_; /**< Map from input keys to internal keys. */
+  dimension_type dimMax_;                                  /**< Maximal dimension of a bar to record. */
+  std::vector<Index_interval> persistenceDiagram_;         /**< Stores current closed persistence intervals. */
+  internal_key numArrow_;                                  /**< Current arrow number. */
+  filtration_value previousFiltrationValue_;               /**< Filtration value of the previous arrow. */
+  /**
+   * @brief filtrationValues_ stores consecutive pairs (i,f) , (j,f') with f != f',
+   * meaning that all inserted faces with key in [i;j-1] have filtration value f,
+   * i is the smallest face index whose face has filtration value f.
+   */
+  std::vector<std::pair<internal_key, filtration_value> > filtrationValues_;
+  Zigzag_persistence<FilteredZigzagOptions, false> pers_; /**< Class computing the pairs. */
+
+  /**
+   * @brief Returns the current persistence diagram without infinite bars.
+   *
+   * @param shortestInterval Intervals shorter than the given value are ignored.
+   * @return Vector of intervals.
+   */
+  std::vector<Filtration_value_interval> _get_persistence_diagram(filtration_value shortestInterval) {
+    std::vector<Filtration_value_interval> diag;
+    diag.reserve(persistenceDiagram_.size());
+
+    std::stable_sort(filtrationValues_.begin(), filtrationValues_.end(),
```

Comment:
Yes, it is. Now, I am very confused about why this line is there. So, I commented it out for now, without removing it, just in case. We sorted the bars by length before, so perhaps it is vestige of it?